fix(express): Ensure 404 routes don't attach route attribute

s1gr1d opened the pull request:
## Which problem is this PR solving?

This is an upstream fix for https://github.com/getsentry/sentry-javascript/pull/16211

Instead of adding `/` to non-existing routes, it sets the attribute to `undefined`, which makes it clear it's a non-existing route and not the index route.

--- 

A tricky thing is that the layers look the same if a non-existing route is requested or the root is requested: `['/', '/']`.
- `/existing` --> `['/', '/existing']`
- `/non-existing` --> `['/', '/']`
- `/` --> `['/', '/']`

So in order to know if it was a 404 route, a utility function that does some additional checking was added.

pichlermarc reviewed (approved):
LGTM! thank you for your contribution @s1gr1d 🙂 

cc @raphael-theriault-swi, @JamieDanielson, @pkanal (component owners)

pichlermarc commented:
oh - looks like there are some lint errors, you can fix them by running `npm run lint:fix` and then pushing the resulting diff 🙂

s1gr1d commented:
I have to look into this again. I see another test is failing. It doesn't add `http.route` either when the route is `/` (but this should be added in this case).

s1gr1d commented:
The problem was that the layers look the same if a non-existing route is requested or the root is requested: ['/', '/']`. 

So in order to know if it was a 404 route, I added a utility function that does some additional checking. I made sure that the existing tests still work and added some more tests for the utility function.

pichlermarc reviewed (commented):
Hi @s1gr1d, thanks for following up on this.

It seems that this PR is based on a commit that's a little over 2 months old. Would you mind rebasing it on top of `main` it so that we can see how this would fit in with the current state? 🙂

david-luna reviewed (approved):
LGTM :) just a couple of nits
